### PR TITLE
Smoke tests - run each suite in its own user-data-dir

### PR DESCRIPTION
**Pull request by lszomoru**
In an effort to improve the noise-to-signal ratio of our smoke tests, I have explored what would it be to run each smoke test suite in its own `--user-data-dir`. This will add roughly a minute to each smoke test run but hopefully solve the random failures that are caused by one tests interfering with another.

### CODE REVIEW

**Review by bpasero — COMMENTED** [review]
I like, thanks 👍 . Some questions:
* I assume this does not break tests that actually need to accumulate UI state (e.g. data loss tests) and that run on stable and insiders build
* can we extract the repeated `setup` and `teardown` code we now have in many tests for some reuse?

**Comment by lszomoru** [comment]
* As far as I can tell the data loss test is always the first smoke test suite that is being executed, so the UI required for this test to pass is accumulated as part of the test. Please correct me if I am wrong if I am missing something. 
* I have reworked the setup/teardown code so that it is being injected for each suite and it only exists in one place.

**Review by bpasero — CHANGES_REQUESTED** [review]
Need to ensure this is not broken:

`yarn smoketest --build /Users/bpasero/Desktop/Visual\ Studio\ Code\ -\ Insiders.app --stable-build /Users/bpasero/Desktop/Visual\ Studio\ Code.app`

File: test/smoke/src/main.ts [hunk 1]
```diff
@@ -299,14 +300,28 @@ describe(`VSCode Smoke Tests (${opts.web ? 'Web' : 'Electron'})`, () => {
...
+				// Add beforeAll hook
+				suite.beforeAll(async function () {
+					const app = new Application(this.defaultOptions);
+					await app!.start(opts.web ? false : undefined);
+					this.app = app;
+				});
+
+				// Move beforeAll to the front of the hook list
+				const suiteBeforeAll = suite['_beforeAll'].pop();
+				suite['_beforeAll'].unshift(suiteBeforeAll);
```
Comment (joaomoreno) [inline test/smoke/src/main.ts:315]: Is this stable enough not to break if we update mocha? Are we allowed to dynamically mess with the suites? Shouldn't we simply find a different way of doing this, like passing along `before` and `after` functions to the `setup*` functions? Eg:

```ts
const before = () => {...};
const after = () => {...};
const opts = { before, after };

setupDataLossTests(opts);
setupDataPreferencesTests(opts);
// ...
```

**Pull request merged**